### ISSUE TITLE
Feature Request: Due Date notifications improvements

### ISSUE THREAD

**Issue by xet7** [wekan/wekan#2567]
Moved to here from https://github.com/wekan/wekan/pull/2536#issuecomment-514819653 that already implements Due Date notifications for Wekan, and is already merged to Wekan master branch.

From @saschafoerster

@whowillcare @xet7 As I was just searching trough all requests just for this feature, maybe this is a helpful collection of open requests related to this pull: 

I was searching for notifications, when dates, especially due dates, where changed. I would like to get this notifications via email, via system notifications (in the card) and via Webhooks of the board. 

- Feature Request : Send notifications of due dates to email, phone etc with selectable notifications to get and settings for it #2297 
- Feature Request: Changing the start & due dates it generates message to Outgoing Webhook #1564 
- Feature Request: Change Due Date and other dates on Multiselect #1410 
- Feature Request: Change key names of Outgoing Webhook #2135 
- Feature Request: Notification Digest Email #2472
- Milestone 36: 2019-07-31 Teams https://github.com/wekan/wekan/milestone/36
- Milestone 34: 2019-09-30 Add/Remove/Bugs https://github.com/wekan/wekan/milestone/34

These open requests are more specific about the email notifications, at some time before the due date, less about the changes of the due date: 

Feature Request: Mails for Due Date (HB)
https://blog.trello.com/custom-due-date-reminders
Would be great, if reminders existed, general option: 1 day before, but also other options like with Trello.
- Feature Request : Send notifications of due dates #2297 
- Feature Request: Send Wekan due notifications to Twilio #2559 
- [Feature] Notification Digest Email #2472 
- Default setting / deadlines notification #2216 
- Milestone 36: 2019-07-31 Teams https://github.com/wekan/wekan/milestone/36

From @whowillcare 

I was thinking about the customizable reminder when I was implementing due
date reminder feature, however, it's little bit more complicated than what
I have implemented.
The trello reminder features are doable but will take later priority, when
I have time.

**Comment by xet7** [comment]
Notification Digest Email

![notification-digest-email](https://user-images.githubusercontent.com/15545/61878371-8b612500-aef9-11e9-8145-1b3888b390d9.png)

**Comment by xet7** [comment]
Moved to here from #2559

Send Wekan due notifications to Twilio

TODO:
- Add docs to wiki, when I get this working. Below is my too complicated way trying to explain settings this up.

Q: how would i connect twilio api into this? so that like let's say when a project is due, i can have it text me?

xet7: Does twilio have incoming webhooks? If not, you could use NODE-RED or some other code to receive wekan board webhook, and then call twilio API. See Wekan GitHub wiki for info about ongoing webhooks. And also on Snap:
```
wekan.help | grep webhook
```
on server shell. Or for Docker, see docker-compose.yml at https://github.com/wekan/wekan
But those due notifications https://github.com/wekan/wekan/pull/2536 are not yet at Snap/Docker, it's only on Wekan source build. Snap and Docker will be fixed soon I presume, I'm testing fixes currently.

Q: I believe twilio has incoming webhooks, https://twilio.com/docs/chat/webhook-events
also, I used snap to install wekan, onto digital ocean vps. not sure if that's relevant.

xet7: In Wekan, snap setting webhooks-attributes defines what info to send to outgoing webhook https://github.com/wekan/wekan-snap/wiki/Supported-settings-keys#list-of-supported-keys
for example,
```
sudo snap set wekan webhooks-attributes='........'
```
And from this pull request code you see what webhooks it maybe has added https://github.com/wekan/wekan/pull/2536 . Then on Wekan board, click `hamburger menu / cog settings icon at right side of members / Outgoing Webhooks`. And then from twilio docs or settings copy incoming webhook URL to Wekan Outgoing Webhook textfield. But I think you still need to wait that Snap version Wekan gets fixed, because it's not fixed yet https://github.com/wekan/wekan/issues/2533 . So that pull request code of due date notifications is not yet in Snap and Docker.

If you need this sooner, at https://github.com/wekan/wekan is `rebuild-wekan.sh` that when running, option 1 installs dependencies and option 2 builds from source, and with `start-wekan.sh` you can start that source built wekan. To `start-wekan.sh`
you need to probably change port and `ROOT_URL`, see https://github.com/wekan/wekan/wiki/Settings
So I don't think for you trying this, it does not require any source code changes to Wekan, it just needs the correct settigns to both Wekan and Twilio. Because you have already used snap, you see your current snap settings with `sudo snap get wekan` . And you can use same settings in
`start-wekan.sh`. If you run source built wekan on same server, you can in
`start-wekan.sh` select different port for wekan node, and same port for mongodb that is in Wekan snap. Hmm, but wekan master branch uses newer Meteor and MongoDB.... so you would could install newer MongoDB 4.0.10 and at /etc/mongod.conf configure MongoDB to run at different port like 27018. Oooops explaining this gets too complicated. I will add new Wekan issue so that I try to make this work with Wekan snap and simpler instructions. Sorry 😃

**Comment by saschafoerster** [comment]
I tested the new notifications with these settings in docker-compose.yml:

     - NOTIFY_DUE_DAYS_BEFORE_AND_AFTER=1
     - NOTIFY_DUE_AT_HOUR_OF_DAY=16

- The card was due on Friday, 02.08.2019, 12:00 (Europe/Berlin)
- Reminder Mail 1 arrives on Friday, 02.08.2019, 18:49 (Europe/Berlin). 
It was expected for Thursday, 01.08.2019, 16:00. 
- Remember Mail 2 arrives on Saturday, 03.08.2019, 18:49 (Europe/Berlin)
It was expected for Friday, 01.08.2019, 12:00. 

Maybe @whowillcare can explain a little more, how I can change the time-zone-settings (maybe by just calculating -2 or -1 to the times in the docker-compose.yml. And I am curious how the corncobs work. Is there a place or option to let them run more often? 
<img width="785" alt="Bildschirmfoto 2019-08-04 um 18 38 26" src="https://user-images.githubusercontent.com/1153741/62427548-eca4a780-b6f4-11e9-9ef3-cf0178a87803.png">
<img width="904" alt="Bildschirmfoto 2019-08-04 um 18 38 36" src="https://user-images.githubusercontent.com/1153741/62427549-ed3d3e00-b6f4-11e9-8b58-f6502adbab8c.png">
<img width="888" alt="Bildschirmfoto 2019-08-04 um 18 38 41" src="https://user-images.githubusercontent.com/1153741/62427550-ed3d3e00-b6f4-11e9-8ab8-bc4cee9597f7.png">

**Comment by saschafoerster** [comment]
Maybe #1696 the timezone setting is related? :)

**Comment by whowillcare** [comment]
This is related to the time zone of where your wekan server resides on (the
host, mostly the linux server).   if it's in UTC, then the calculation will
be way off than user's actual time zone.

**Comment by saschafoerster** [comment]
@whowillcare What about docker installations? Is there a way to adjust the timing in the container?

**Comment by adelphi-company** [comment]
Wekan 3.23 (snap):
It seems that the "due date notification" ignores the "change watch: mute" setting.
All notifications are getting suppressed except "due" notifications. Was this on purpose?

I also notices that a notification is send out to ALL (!) board members when adding a due to a card, even if the due is beyond default limit (NOTIFY_DUE_DAYS_BEFORE_AND_AFTER=2). Was this on purpose? What about sending notifications only to the members that are assigned to the modified card?

Is there any option to control these behaviours individually for each board or maybe globally ?
My users are complainig about too much emails coming from wekan...

**Comment by xet7** [comment]
@adelphi-company 

Then change your settings:
https://github.com/wekan/wekan/blob/master/docker-compose.yml#L248-L275

Same in Snap:
```
sudo snap set wekan bigevents-pattern='.....'
```

**Comment by saschafoerster** [comment]
> # Notify due days, default 2 days before and after. 0 = due notifications disabled. Default: 2
>       #- NOTIFY_DUE_DAYS_BEFORE_AND_AFTER=2

I would like to only get a single due notification, not one before and one after. It actually only fills up our mailboxes a second time, so we always have to check a second time, if something has been done. I would love to see an option to only get a due notification a certain amount of time before, ideally individually configurable per card.

**Comment by saschafoerster** [comment]
Wekan is sending due-date notifications for cards in archive. I think archived cards shouldn't send notifications.

**Comment by whowillcare** [comment]
will take a look, and will be fixing this issue and adding new feature
about due notification soon.

**Comment by saschafoerster** [comment]
I changed the configuration back to 
      - NOTIFY_DUE_DAYS_BEFORE_AND_AFTER=2,0
      - NOTIFY_DUE_AT_HOUR_OF_DAY=8
      - EMAIL_NOTIFICATION_TIMEOUT=30000

Now we got due-notification-emails again. When I create new test-entries, they were all sent correctly and as expected. 

For other cards in productive boards we had some problems: sometimes names in the e-mails where not correct, sometimes they notifications where sent to the wrong people. 

For debugging I need to know, how this notifications are created: is this a cronjob, which checks all cards, or do notifications stop working, when updating wekan and have to be reset after a restart?
